Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs

```diff
@@ -0,0 +1,522 @@
+                    symbol.SecurityType != _symbolMapper.GetLeanSecurityType(symbol.Value))
+                {
+                    continue;
+                }
+
+                WebSocket.Send(JsonConvert.SerializeObject(new
+                {
+                    @event = "subscribe",
+                    channel = "book",
+                    pair = _symbolMapper.GetBrokerageSymbol(symbol)
+                }));
+            }
+
+            Log.Trace("BitfinexBrokerage.Subscribe: Sent subscribe.");
+        }
+
+        /// <summary>
+        /// Ends current subscriptions
+        /// </summary>
+        public void Unsubscribe(IEnumerable<Symbol> symbols)
+        {
+            if (WebSocket.IsOpen)
+            {
+                var map = ChannelList.ToDictionary(k => k.Value.Symbol, k => k.Key, StringComparer.InvariantCultureIgnoreCase);
+                foreach (var symbol in symbols)
+                {
+                    if (map.ContainsKey(symbol.Value))
+                    {
+                        WebSocket.Send(JsonConvert.SerializeObject(new
+                        {
+                            @event = "unsubscribe",
+                            channelId = map[symbol.Value]
+                        }));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Implementation of the OnMessage event
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void OnMessageImpl(object sender, WebSocketMessage e)
+        {
+            try
+            {
+                var token = JToken.Parse(e.Message);
+
+                if (token is JArray)
+                {
+                    int channel = token[0].ToObject<int>();
+                    if (channel != 0 && token[1].Type != JTokenType.String)
+                    {
+                        if (token.Count() == 2)
+                        {
+                            OnSnapshot(
+                                token[0].ToObject<string>(),
+                                token[1].ToObject<string[][]>()
+                            );
+                        }
+                        else
+                        {
+                            OnUpdate(
+                                token[0].ToObject<string>(),
+                                token.ToObject<string[]>().Skip(1).ToArray()
+                            );
+                        }
+                    }
+                    else if (channel == 0)
+                    {
+                        string term = token[1].ToObject<string>();
+                        switch (term.ToLower())
+                        {
+                            case "oc":
+                                OnOrderClose(token[2].ToObject<string[]>());
+                                return;
+                            case "tu":
+                                OnTradeExecuted(token[2].ToObject<string[]>());
```

Comment:
Is `channel == 0` only for our trades? Do we need to handle `"tu"` messages also if `channel != 0` ?

Reply:
Yeah, channel == 0 only for our trades.

>Do we need to handle "tu" messages also if channel != 0
If we want to we need to subscribe to trade channel of specific symbol. Do we want to emit trade tick or all subscribed symbols or only for specific order?